coupons/forms.py

```diff
@@ -53,15 +53,14 @@ def clean_code(self):
 
         try:  # check if there is a user bound coupon existing
             user_coupon = coupon.users.get(user=self.user)
-            if user_coupon.redeemed_at is not None:
-                raise forms.ValidationError(_("This code has already been used by your account."))
+            if user_coupon.fully_redeemed:
+                raise forms.ValidationError(
+                    _("This code has already been fully used by your account."))
         except CouponUser.DoesNotExist:
-            if coupon.user_limit is not 0:  # zero means no limit of user count
```

Comment:
All these checks are good at the form clean level, but they don't protect against race conditions on save.  1) The save occurs later, and 2)  Even if this code was just prior to save(), it itself is subject to race conditions (any read-check-write cycle is).  I'm pretty sure that we can get some bad coupon use situations into the DB currently.  Protecting against this would require either the save() to use a select_for_update, or DB level constaints (by far the safest) or only only using update() with F() conditions...

Reply:
Updated to use `select_for_update` with a message as to how `select_for_update` will have no practical effect as `autocommit` will release the lock after each query unless the whole thing is explicitly wrapped in an atomic transaction block.

Reply:
Using select_for_update inside transaction here: https://github.com/Tesloop/tesloop/pull/237/files#diff-8803c135ecd261fcb3f3597220fa4b09R61